text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,626 @@
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  PM1[Publish m1]
+  end
+  subgraph CONSUMER
+  DM1[Deliver m1]
+  end
+  PM1-. link .->DM1;
+  PM1-->DM1;
+
+  classDef normal fill:green
+  class PM1,DM1 normal
+  linkStyle 0,1 color:green,stroke:green
+```
+
+It is recommended to add spans for settlement operations. Those spans can
+either be created manually or via auto-instrumentation:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+    direction TB
+    PM1[Publish m1]
+  end
+  subgraph CONSUMER
+    direction TB
+    DM1[Deliver m1]-->S1[Settle m1]
+  end
+  PM1-. link .->DM1;
+  PM1-. link .->S1;
+
+  classDef normal fill:green
+  class PM1,DM1,S1 normal
+  linkStyle 0,1,2 color:green,stroke:green
+```
+
+A producer publishes a batch of messages, single messages are delivered to consumers:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  P[Publish]
+  end
+  subgraph CONSUMER
+  direction TB
+  DM1[Deliver m1]
+  DM2[Deliver m2]
+  end
+  P-. link .->DM1;
+  P-. link .->DM2;
+
+  classDef normal fill:green
+  class P,DM1,DM2 normal
+  linkStyle 0,1 color:green,stroke:green
+```
+
+When consuming a single message, the "Deliver" spans can be parented to the creation context:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  P[Publish]
+  end
+  subgraph CONSUMER
+  direction TB
+  DM1[Deliver m1]
+  DM2[Deliver m2]
+  end
+  P-. link .->DM1;
+  P-. link .->DM2;
+  P-->DM1;
+  P-->DM2;
+
+  classDef normal fill:green
+  class P,DM1,DM2 normal
+  linkStyle 0,1,2,3 color:green,stroke:green
+```
+
+A producer creates and publishes a single message, it is delivered as part of a
+batch of message to a consumer:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  PM1[Publish m1]
+  PM2[Publish m2]
+  end
+  subgraph CONSUMER
+  direction TB
+  D[Deliver]-.-PRM1[Process m1]
+  D-.-PRM2[Process m2]
+  end
+  PM1-. link .->D;
+  PM2-. link .->D;
+
+  classDef normal fill:green
+  class PM1,PM2,D normal
+  classDef additional opacity:0.4
+  class PRM1,PRM2 additional
+  linkStyle 0,1 opacity:0.4
+  linkStyle 2,3 color:green,stroke:green
+```
+
+### Pull-based scenarios
+
+A producer creates and publishes a single message, the single message is
+delivered to a consumer. "Create" spans are created independently of the
+"Publish" operation:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  A[Ambient]-->CM1[Create m1]
+  A-->CM2[Create m2]
+  A-->P[Publish]
+  end
+  subgraph CONSUMER
+  direction TB
+  RM1[Receive m1]
+  RM2[Receive m2]
+  end
+  CM1-. link .->RM1;
+  CM2-. link .->RM2;
+  CM1-. link .->P;
+  CM2-. link .->P;
+
+  classDef normal fill:green
+  class CM1,CM2,P,RM1,RM2 normal
+  classDef additional opacity:0.4
+  class A additional
+  linkStyle 0,1,2 opacity:0.4
+  linkStyle 3,4,5,6 color:green,stroke:green
+```
+
+"Create" spans are created as part of the "Publish" operation:
```

Comment:
in the description above, publish span above needs to link to create spans, so it seems those should be created first and then publish spans has to link to them. Then they should be siblings (if all happen in the same ambient context)

Reply:
Isn't that exactly what's depicted in the diagram above? That's the header:

> A producer creates and publishes a single message, the single message is delivered to a consumer. "Create" spans are created independently of the "Publish" operation:

And that's the diagram:
![image](https://user-images.githubusercontent.com/16337442/231582674-403935d9-41f9-4eb2-808e-102479eebc68.png)